### ISSUE TITLE
Compiler aborting due to parser-detected syntax error

### ISSUE THREAD

**Issue by khfeng** [acpica/acpica#756]
```
$ iasl -tc dsdt.dsl
...
Compiler aborting due to parser-detected syntax error(s)
dsdt.dsl    850:                 GSWS (0x91F3)
Error    6084 -                    ^ Object does not exist (GSWS)
``

However, GSWS can be found in DSDT:
```
    Scope (\)
    {
        Method (GSWS, 1, Serialized)
        ...
    }
```

**Comment by khfeng** [comment]
[dump.txt](https://github.com/acpica/acpica/files/8183480/dump.txt)
Full ACPI dump

**Comment by acpibob** [comment]
I will take a look at this when I can.

**Comment by acpibob** [comment]
dsdt.dsl    850:                 GSWS (0x91F3)
 Well, your code has a reference to GSWS at line 850, and

GSWS is defined on this line:

 41981:       Method (GSWS, 1, Serialized)
        {

So, the invocation of GSWS on line 850 appears to be an illegal forward reference.  However, the compiler should flag this with a more appropriate error message.

**Comment by khfeng** [comment]
Is it possible to make it work? Apparently the ASL can be compiled just fine under Windows.

**Comment by acpibob** [comment]
You can try using the -f option to iASL. This causes a "force generate" of the .AML file.

When you mention that "can be compiled just fine under Windows" do you mean using the MS ASL compiler? If so, this compiler lets errors through, until the code is actually executed, then it fails.

**Comment by khfeng** [comment]
I am just assuming ODM/OEM is using Windows to generate the BIOS, I don't really know what's the compiler they are using.

Is forward reference not allowed by the spec?

**Comment by khfeng** [comment]
Thanks for the clarification. Let me forward this info to the BIOS folks.

**Comment by khfeng** [comment]
Is it possible to enable forward reference as an option? Otherwise I can't patch DSDT/SSDT for debugging purpose.

**Comment by acpibob** [comment]
No, it is much too complicated to make this an option. BTW, this will fail on Windows.

**Comment by acpibob** [comment]
I'm seeing errors like this one:

Compiler aborting due to parser-detected syntax error(s)
dsdt.dsl   4545:                             ElseIf ((_T_1 == 0x56))
Error    6126 -                                  ^ syntax error, unexpected PARSEOP_ELSEIF

Which is caused by unresolved externals

**Comment by khfeng** [comment]
These errors on ADL platforms make debugging extremely hard. Many issues require assistance from DSDT/SSDT overlay.

**Comment by acpibob** [comment]
Here is a fixed DSDT that compiles (albeit with lots of warnings and remarks).  The problem was the definition of control methods (which cannot be disassembled correctly without an External statement embedded in the AML). File renamed from dsdt.dsl to dsdt.txt:

https://github.com/acpica/acpica/files/8444829/dsdt.txt
